### PR TITLE
Implement ZCash address discovery

### PR DESCRIPTION
**Pull request by cypt4**
Resolves https://github.com/brave/brave-browser/issues/33662

Resolves 

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/brave_wallet/browser/zcash/zcash_grpc_utils.cc [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+#include "base/logging.h"
+
+namespace brave_wallet {
+
+namespace {
+constexpr size_t kMaxMessageSizeBytes = 10000;
+}  // namespace
+
+std::string GetPrefixedProtobuf(const std::string& serialized_proto) {
+  char compression = 0;  // 0 means no compression
```
[the remark this answers is not in the document]
Comment (fmarier) [reply components/brave_wallet/browser/zcash/zcash_grpc_utils.cc:21]: False positive: https://github.com/brave/security-action/issues/382

**Review by nuo-xu — APPROVED** [review]
iOS++

File: components/brave_wallet/browser/zcash/zcash_wallet_service_tasks.cc [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+        base::BindOnce(&DiscoverNextUnusedZCashAddressTask::OnGetLastBlock,
+                       this));
+    return;
+  }
+
+  zcash_wallet_service_->zcash_rpc()->IsKnownAddress(
+      GetNetworkForZCashKeyring(account_id_->keyring_id),
+      current_address_->address_string, 1, block_end_.value(),
+      base::BindOnce(&DiscoverNextUnusedZCashAddressTask::OnGetIsKnownAddress,
+                     this));
```
[the remark this answers is not in the document]
Comment (cypt4) [reply components/brave_wallet/browser/zcash/zcash_wallet_service_tasks.cc:123]: Refcounted

File: components/brave_wallet/browser/zcash/zcash_wallet_service_tasks.cc [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+    error_ = "Internal error";
+    ScheduleWorkOnTask();
+    return;
+  }
+
+  if (!block_end_) {
+    zcash_wallet_service_->zcash_rpc()->GetLatestBlock(
+        GetNetworkForZCashKeyring(account_id_->keyring_id),
+        base::BindOnce(&DiscoverNextUnusedZCashAddressTask::OnGetLastBlock,
+                       this));
```
[the remark this answers is not in the document]
Comment (cypt4) [reply components/brave_wallet/browser/zcash/zcash_wallet_service_tasks.cc:115]: Refcounted

File: components/brave_wallet/browser/zcash/zcash_wallet_service_tasks.cc [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+      start_address_(std::move(start_address)),
+      callback_(std::move(callback)) {}
+
+DiscoverNextUnusedZCashAddressTask::~DiscoverNextUnusedZCashAddressTask() =
+    default;
+
+void DiscoverNextUnusedZCashAddressTask::ScheduleWorkOnTask() {
+  base::SequencedTaskRunner::GetCurrentDefault()->PostTask(
+      FROM_HERE,
+      base::BindOnce(&DiscoverNextUnusedZCashAddressTask::WorkOnTask, this));
```
[the remark this answers is not in the document]
Comment (cypt4) [reply components/brave_wallet/browser/zcash/zcash_wallet_service_tasks.cc:65]: Refcounted

**Comment by cypt4** [comment]
needs-security-label removed since feature is under the flag. And will have separate sec-review before the release

File: components/brave_wallet/browser/zcash/zcash_wallet_service_tasks.h [hunk 5]
```diff
@@ -0,0 +1,120 @@
+/* Copyright (c) 2023 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_COMPONENTS_BRAVE_WALLET_BROWSER_ZCASH_ZCASH_WALLET_SERVICE_TASKS_H_
+#define BRAVE_COMPONENTS_BRAVE_WALLET_BROWSER_ZCASH_ZCASH_WALLET_SERVICE_TASKS_H_
+
+#include <algorithm>
```
Comment (supermassive) [inline components/brave_wallet/browser/zcash/zcash_wallet_service_tasks.h:9]: ~#include <algorithm>~

File: components/brave_wallet/browser/zcash/zcash_grpc_utils.cc [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+  char buff[4];          // big-endian 4 bytes of message size
+  base::WriteBigEndian<uint32_t>(buff, serialized_proto.size());
+  std::string result;
+  result.append(&compression, 1);
+  result.append(buff, 4);
+  result.append(serialized_proto);
+  return result;
+}
```
Comment (supermassive) [inline components/brave_wallet/browser/zcash/zcash_grpc_utils.cc:30]:
```
  std::string result(kGrpcHeaderSize);
  result[0] = 0; // no compression
  base::WriteBigEndian<uint32_t>(result[1], serialized_proto.size());

  result.append(serialized_proto);
  return result;
```

File: components/brave_wallet/browser/zcash/zcash_grpc_utils.cc [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+void GRrpcMessageStreamHandler::OnDataReceived(std::string_view string_piece,
+                                               base::OnceClosure resume) {
+  data_.append(string_piece);
+  std::string_view data(data_);
+  size_t pos = 0;
+
+  bool should_resume = false;
+  while (!should_resume) {
+    if (!data_to_read_ && data.size() > kGrpcHeaderSize) {
+      if (data[0] != 0) {
```
Comment (supermassive) [inline components/brave_wallet/browser/zcash/zcash_grpc_utils.cc:65]: 0 needs to be replaced everywhere with constant `kNoCompression`

File: components/brave_wallet/browser/zcash/zcash_grpc_utils.cc [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+        OnComplete(false);
+        return;
+      }
+      uint32_t size = 0;
+      base::ReadBigEndian(reinterpret_cast<const uint8_t*>(&(data[1])), &size);
+      data_to_read_ = size;
```
Comment (supermassive) [inline components/brave_wallet/browser/zcash/zcash_grpc_utils.cc:71]: We don't need data_to_read_ out of this scope. Just parse these bytes one evey iteration

File: components/brave_wallet/browser/zcash/zcash_grpc_utils.cc [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+      }
+
+      data = data.substr(kGrpcHeaderSize + *data_to_read_);
+      pos += kGrpcHeaderSize + *data_to_read_;
+      data_to_read_.reset();
+    } else {
+      should_resume = true;
+    }
+  }
+  data_ = data_.substr(pos);
```
Comment (supermassive) [inline components/brave_wallet/browser/zcash/zcash_grpc_utils.cc:92]: We don't need extra pos variable. 
`data_ = std::string(data_view);` should be enough

File: components/brave_wallet/browser/zcash/zcash_grpc_utils.cc [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+      data_to_read_ = size;
+      if (*data_to_read_ > kMaxMessageSizeBytes) {
+        // Too large message
+        OnComplete(false);
+        return;
+      }
+    }
+
+    if (data_to_read_ && data.size() >= (kGrpcHeaderSize + *data_to_read_)) {
+      if (!ProcessMessage(data.substr(0, kGrpcHeaderSize + *data_to_read_))) {
```
Comment (supermassive) [inline components/brave_wallet/browser/zcash/zcash_grpc_utils.cc:80]: Is header part of message?
should that be `data.substr(kGrpcHeaderSize, *data_to_read_)` ?

File: components/brave_wallet/browser/zcash/zcash_grpc_utils.cc [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_wallet/browser/zcash/zcash_grpc_utils.h"
+
+#include <utility>
+
+#include "base/big_endian.h"
+#include "base/functional/callback.h"
+#include "base/logging.h"
```
Comment (supermassive) [inline components/brave_wallet/browser/zcash/zcash_grpc_utils.cc:12]: ~`#include "base/logging.h"`~

File: components/brave_wallet/browser/zcash/zcash_grpc_utils.cc (hunk 10 above)
Comment (cypt4) [reply components/brave_wallet/browser/zcash/zcash_grpc_utils.cc:80]: Yes it is

**Review by supermassive — APPROVED** [review]

File: components/brave_wallet/browser/zcash/zcash_wallet_service_tasks.h [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+#include <string>
+#include <vector>
+
+#include "brave/components/brave_wallet/browser/zcash/zcash_wallet_service.h"
+#include "brave/components/brave_wallet/common/brave_wallet.mojom.h"
+
+namespace brave_wallet {
+
+class GetTransparentUtxosContext
+    : public base::RefCountedThreadSafe<GetTransparentUtxosContext> {
```
Comment (bridiver) [inline components/brave_wallet/browser/zcash/zcash_wallet_service_tasks.h:19]: why are these RefCountedThreadSafe? Use of ref-counting should be limited to cases where something is truly shared and also this class does not appear to be thread safe 
https://www.chromium.org/developers/coding-style/important-abstractions-and-data-structures/#scoped_refptrt-baserefcounted-baserefcountedthreadsafe

File: components/brave_wallet/browser/zcash/zcash_wallet_service_tasks.cc [hunk 13]
```diff
@@ -0,0 +1,343 @@
...
+  CHECK(keyring_service);
+  auto next_key_id = current_address_->key_id.Clone();
+
+  next_key_id->index++;
+
+  return keyring_service->GetZCashAddress(*account_id_, *next_key_id);
+}
+
+void DiscoverNextUnusedZCashAddressTask::WorkOnTask() {
+  if (!callback_) {
```
Comment (bridiver) [inline components/brave_wallet/browser/zcash/zcash_wallet_service_tasks.cc:81]: if this is passed an empty callback it will leak

**Pull request merged**